features/lorawan/LoRaWANStack.cpp

```diff
@@ -492,322 +363,526 @@ int16_t LoRaWANStack::handle_rx(uint8_t* data, uint16_t length, uint8_t& port, i
-                _rx_msg.msg.mcps_indication.buffer = mcps_indication->buffer;
-                _rx_msg.msg.mcps_indication.type = mcps_indication->type;
-
-                // Notify application about received frame..
-                tr_debug("Received %d bytes", _rx_msg.msg.mcps_indication.buffer_size);
-                _rx_msg.receive_ready = true;
-
-                send_event_to_application(RX_DONE);
-
-                /*
-                 * If fPending bit is set we try to generate an empty packet
-                 * with CONFIRMED flag set. We always set a CONFIRMED flag so
-                 * that we could retry a certain number of times if the uplink
-                 * failed for some reason
-                 * or
-                 * Class C and node received a confirmed message so we need to
-                 * send an empty packet to acknowledge the message.
-                 * This scenario is unspecified by LoRaWAN 1.0.2 specification,
-                 * but version 1.1.0 says that network SHALL not send any new
-                 * confirmed messages until ack has been sent
-                 */
-                if ((_loramac.get_device_class() != CLASS_C && mcps_indication->fpending_status) ||
-                    (_loramac.get_device_class() == CLASS_C && mcps_indication->type == MCPS_CONFIRMED)) {
-#if (MBED_CONF_LORA_AUTOMATIC_UPLINK_MESSAGE)
-                    tr_debug("Sending empty uplink message...");
-                    send_automatic_uplink_message(mcps_indication->port);
-#else
-                    send_event_to_application(UPLINK_REQUIRED);
-#endif
-                }
-            } else {
-                // Invalid port, ports 0, 224 and 225-255 are reserved.
-            }
-            break;
+
+        // Class A unconfirmed message sent, TX_DONE event will be sent to
+        // application when RX2 windows is elapsed, i.e., in process_reception_timeout()
+        _ctrl_flags &= ~LORAMAC_TX_ONGOING;
+        _ctrl_flags |= LORAMAC_TX_DONE;
+
+        // In Class C, reception timeout never happens, so we handle the state
+        // progression here
+        if (_loramac.get_device_class() == CLASS_C) {
+            _loramac.post_process_mcps_req();
+            state_controller(DEVICE_STATE_STATUS_CHECK);
+            state_machine_run_to_completion();
         }
     }
 }
 
-lorawan_status_t LoRaWANStack::set_link_check_request()
+void LoRaWANStack::process_reception(const uint8_t* const payload, uint16_t size,
+                                     int16_t rssi, int8_t snr)
 {
-    _link_check_requested = true;
-    if (!_callbacks.link_check_resp) {
-        tr_error("Must assign a callback function for link check request. ");
-        return LORAWAN_STATUS_PARAMETER_INVALID;
+    _device_current_state = DEVICE_STATE_RECEIVING;
+
+    _loramac.on_radio_rx_done();
+
+    loramac_mhdr_t mac_hdr;
+    uint8_t pos = 0;
+    mac_hdr.value = payload[pos++];
+
+    /**
```

Comment:
I'm still not really grasping why all this MAC functionality is being moved out of the MAC. This totally stops the MAC actually being, well, a MAC. You've now got two "stack" and "MAC" classes which actually form the MAC between them, and I'm not immediately seeing where the dividing line is conceptually.

Maybe there is no clear line, but in that case the MAC class could be eliminated and just merge with this.

This is also less robust and clear than the original MAC code it's replacing - this lost the `switch` and is letting through reserved packets as a result.